features/nvstore/source/nvstore.h

```diff
@@ -0,0 +1,306 @@
+ *        Warning: This function is not thread safe and should not be called
+ *        concurrently with other NVStore functions.
+ *
+ * @returns NVSTORE_SUCCESS       Deinitialization completed successfully.
+ */
+    int deinit();
+
+/**
+ * @brief Reset Flash NVStore areas.
+ *        Warning: This function is not thread safe and should not be called
+ *        concurrently with other NVStore functions.
+ *
+ * @returns NVSTORE_SUCCESS       Reset completed successfully.
+ *          NVSTORE_READ_ERROR    Physical error reading data.
+ *          NVSTORE_WRITE_ERROR   Physical error writing data.
+ */
+    int reset();
+
+/**
+ * @brief Return NVStore size (area size).
+ *
+ * @returns NVStore size.
+ */
+    uint32_t size();
+
+
+/**
+ * @brief Return address and size of an NVStore area.
+ *
+ * @returns NVSTORE_SUCCESS           Success.
+ *          NVSTORE_BAD_VALUE         Bad area parameter.
+ */
+    int get_area_params(uint8_t area, uint32_t &address, size_t &size);
+
+
+/**
+ * @brief Returns one item of data programmed on Flash, given key.
+ *        This is a self contained version of the get function (not requiring init), traversing the flash each time if triggered.
+ *        This function is NOT thread safe. Its implementation is here for the case we want to minimise code size for clients
+ *        such as boot loaders, performing minimal accesses to NVstore. In this case all other APIs can be commented out.
+ *
+ * @param[in]  key                  Key of stored item (must be between 0-15).
+ *
+ * @param[in]  buf_len_bytes        Length of input buffer in bytes.
+ *
+ * @param[in]  buf                  Buffer to store data on (must be aligned to a 32 bit boundary).
+ *
+ * @param[out] actual_len_bytes     Actual length of returned data
+ *
+ * @returns NVSTORE_SUCCESS           Value was found on Flash.
+ *          NVSTORE_NOT_FOUND         Value was not found on Flash.
+ *          NVSTORE_READ_ERROR        Physical error reading data.
+ *          NVSTORE_DATA_CORRUPT      Data on Flash is corrupt.
+ *          NVSTORE_BAD_VALUE         Bad value in any of the parameters.
+ *          NVSTORE_BUFF_TOO_SMALL    Not enough memory in user buffer.
+ *          NVSTORE_BUFF_NOT_ALIGNED  Buffer not aligned to 32 bits.
+ */
+    int probe(uint16_t key, uint16_t buf_len_bytes, uint32_t *buf, uint16_t &actual_len_bytes);
+
+private:
+    typedef struct
+    {
+        uint32_t address;
+        size_t   size;
+    } nvstore_area_data_t;
+
+    int _init_done;
+    uint32_t _init_attempts;
+    uint8_t _active_area;
+    uint16_t _max_keys;
+    uint16_t _active_area_version;
+    uint32_t _free_space_offset;
+    uint32_t _size;
+    NVstoreSharedLock _lock;
+    uint32_t *_offset_by_key;
+    nvstore_area_data_t _flash_area_params[NVSTORE_NUM_AREAS];
+
+    // Private constructor, as class is a singleton
+    NVStore();
+
+    int flash_read_area(uint8_t area, uint32_t offset, uint32_t len_bytes, uint32_t *buf);
```

Comment:
Can we get some documentation on private functions? Dont need to expose through doxy, just explanation of what and why they exist.

Reply:
Will add.